### PR TITLE
KAFKA-4218: Enable access to key in ValueTransformer

### PR DESCRIPTION
**Pull request by jeyhunkarimov**
This PR groups KAFKA-4218,  KAFKA-4726 and KAFKA-3745.

### CODE REVIEW

**Comment by mjsax** [comment]
Thanks for the PR.
As this is a public API change, we need a KIP...

**Comment by jeyhunkarimov** [comment]
Thanks for reminding, done.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java [hunk 1]
```diff
@@ -79,7 +79,7 @@
      * @param value the value to be transformed
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: The Javadoc should also explain the new key param.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java [hunk 2]
```diff
@@ -35,13 +35,13 @@
  * @see KTable#mapValues(ValueMapper)
  */
 @InterfaceStability.Unstable
-public interface ValueMapper<V, VR> {
+public interface ValueMapper<K, V, VR> {
 
     /**
      * Map the given value to a new value.
      *
      * @param value the value to be mapped
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java]: The Javadoc should also explain the new key param.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java [hunk 3]
```diff
@@ -35,13 +35,13 @@
  * @see KTable#mapValues(ValueMapper)
  */
 @InterfaceStability.Unstable
-public interface ValueMapper<V, VR> {
+public interface ValueMapper<K, V, VR> {
 
     /**
      * Map the given value to a new value.
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapper.java]: The 1-liner description should be updated to match the new (key, val) -> newVal behavior.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java (hunk 1 above)
Comment (miguno) [reply streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: Also , the 1-liner description "Transform the given value to a new value." needs updating.

**Comment by miguno** [comment]
Thanks for this PR and the associated [KIP-149](https://cwiki.apache.org/confluence/display/KAFKA/KIP-149%3A+Enabling+key+access+in+ValueTransformer%2C+ValueMapper%2C+and+ValueJoiner), @jeyhunkarimov!

(Review ongoing)

**Comment by jeyhunkarimov** [comment]
Thanks for review @miguno. Because my initial solution didn't guard the key object and was not backwards-compatible (as discussed in mailing list), I had to change the PR and KIP accordingly.

**Pull request closed without merging**